src/parameterFiles.ts

```diff
@@ -0,0 +1,507 @@
+  sortQuickPickList(pickItems);
+
+  // Move the current item (if any) to the top
+  const currentItem = pickItems.find(pi => pi.data === currentParamFile);
+  if (currentItem) {
+    // tslint:disable-next-line: no-any
+    pickItems = [currentItem].concat(pickItems.filter(ppf => ppf !== currentItem));
+  }
+
+  // Add None at top, Browse/Open Current at bottom
+  const none: IAzureQuickPickItem<IPossibleParamFile | undefined> = {
+    label: "$(circle-slash) None",
+    description: !!currentParamUri ? undefined : currentMessage,
+    data: undefined
+  };
+  const browse: IAzureQuickPickItem<IPossibleParamFile | undefined> = {
+    label: '$(file-directory) Browse...',
+    data: undefined
+  };
+  const openCurrent: IAzureQuickPickItem<IPossibleParamFile | undefined> = {
+    label: '$(split-horizontal) Open Current',
+    data: undefined
+  };
+  pickItems = [none].concat(pickItems).concat([browse]);
+
+  if (currentItem) {
+    pickItems = pickItems.concat([openCurrent]);
+  }
+
+  return {
+    items: pickItems,
+    currentParamFile,
+    none,
+    browse,
+    openCurrent
+  };
+}
+
+function sortQuickPickList(pickItems: IAzureQuickPickItem<IPossibleParamFile | undefined>[]): void {
+  pickItems.sort((a, b) => {
+    const aData = a?.data;
+    const bData = a?.data;
+
+    // Close name matches go first
+    if (a?.data?.isCloseNameMatch !== b?.data?.isCloseNameMatch) {
+      return a?.data?.isCloseNameMatch ? -1 : 1;
+    }
+
+    // Otherwise compare filenames
+    // tslint:disable-next-line: strict-boolean-expressions
+    return (aData?.uri.fsPath || "").localeCompare(bData?.uri.fsPath || "");
+  });
+}
+
+function createQuickPickItem(paramFile: IPossibleParamFile, current: IPossibleParamFile | undefined, templateUri: Uri): IAzureQuickPickItem<IPossibleParamFile> {
+  const isCurrent: boolean = paramFile === current;
+  return {
+    label: `${isCurrent ? '$(check) ' : '$(json) '} ${paramFile.friendlyPath}`,
+    data: paramFile,
+    description: isCurrent ? currentMessage
+      : paramFile.isCloseNameMatch ? similarFilenameMessage
+        : undefined
+  };
+}
+
+function normalizePath(filePath: Uri | string): string {
+  const fsPath: string = typeof filePath === 'string' ? filePath :
+    filePath.fsPath;
+  let normalizedPath = path.normalize(fsPath);
+  if (isWin32) {
+    normalizedPath = normalizedPath.toLowerCase();
+  }
+
+  return normalizedPath;
+}
+
+/**
+ * Finds parameter files to suggest for a given template.
+ */
+export async function findSuggestedParameterFiles(templateUri: Uri): Promise<IPossibleParamFile[]> {
+  let paths: IPossibleParamFile[] = [];
+
+  // Current logic is simple: Find all .json/c files in the same folder as the template file and check
+  //   if they're a parameter file
+  try {
+    const folder = path.dirname(templateUri.fsPath);
+    const fileNames: string[] = await fse.readdir(folder);
+    for (let paramFileName of fileNames) {
```

Comment:
>for (let paramFileName of fileNames)  [](start = 4, length = 37)

Would it make sense to log in telemetry length of time it took us to search all files? Mostly concerned that some people (and they will be those users :) ) may have thousands of files in one folder.

Reply:
And if we did see long delays, then maybe be more selective about naming patterns. Aren't they typically .parameters.json? 

---
In reply to: [389206328](https://github.com/microsoft/vscode-azurearmtools/pull/514#discussion_r389206328) [](ancestors = 389206328)

Reply:
Not a bad idea.  Common patterns are .params and .parameters, but the could be anything.

Reply:
Wrapped with callWithTelemetryAndErrorHandling, which will automatically add timing.